### PR TITLE
BUG: json invoke default handler for unsupported numpy dtypes

### LINKED ISSUE #12554: to_json raises 'Unhandled numpy dtype 15' with complex values.
#### Code Sample, a copy-pastable example if possible
import pandas
df = pandas.DataFrame({'a': [complex(3, -1)]})
df.to_json()

#### Expected Output
'{"a":{"0":3-1j}}'

#### Actual output:
```
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-3-b53888fd1707> in <module>()
----> 1 df.to_json()

.../lib/python2.7/site-packages/pandas/core/generic.pyc in to_json(self, path_or_buf, orient, date_format, double_precision, force_ascii, date_unit, default_handler)
    892             force_ascii=force_ascii,
    893             date_unit=date_unit,
--> 894             default_handler=default_handler)
    895 
    896     def to_hdf(self, path_or_buf, key, **kwargs):

.../lib/python2.7/site-packages/pandas/io/json.pyc in to_json(path_or_buf, obj, orient, date_format, double_precision, force_ascii, date_unit, default_handler)
     33             obj, orient=orient, date_format=date_format,
     34             double_precision=double_precision, ensure_ascii=force_ascii,
---> 35             date_unit=date_unit, default_handler=default_handler).write()
     36     else:
     37         raise NotImplementedError("'obj' should be a Series or a DataFrame")

.../lib/python2.7/site-packages/pandas/io/json.pyc in write(self)
     76             date_unit=self.date_unit,
     77             iso_dates=self.date_format == 'iso',
---> 78             default_handler=self.default_handler)
     79 
     80 

RuntimeError: Unhandled numpy dtype 15
```

#### output of ``pd.show_versions()``

```
INSTALLED VERSIONS
------------------
commit: None
python: 2.7.11.final.0
python-bits: 64
OS: Darwin
OS-release: 15.0.0
machine: x86_64
processor: i386
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8

pandas: 0.17.1
nose: 1.3.7
pip: 8.0.3
setuptools: 19.7
Cython: 0.23.4
numpy: 1.8.2
scipy: 0.15.1
statsmodels: 0.6.1
IPython: 4.0.1
sphinx: 1.3.3
patsy: 0.4.1
dateutil: 2.4.2
pytz: 2015.7
blosc: None
bottleneck: None
tables: None
numexpr: None
matplotlib: 1.4.3
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: None
httplib2: None
apiclient: None
sqlalchemy: None
pymysql: None
psycopg2: None
Jinja2: None
```

### PR DESCRIPTION
**Pull request by Komnomnomnom**
- [x] closes #12554
- [x] tests added / passed
- [x] passes `git diff upstream/master | flake8 --diff`
- [x] whatsnew entry
- [x] valgrind clean
- [ ] windows tests ok
- [x] vbench ok

Recommend to merge after #12802 has been accepted. (valgrind should be clean then)

### CODE REVIEW

File: doc/source/whatsnew/v0.18.1.txt [hunk 1]
```diff
@@ -203,6 +203,7 @@ Performance Improvements
 
 
 
+- Bug in ``DataFrame.to_json`` with unsupported `dtype` not passed to default handler (:issue:`12554`).
```
Comment (jreback) [inline doc/source/whatsnew/v0.18.1.txt]: should go in bug fix section

File: pandas/io/tests/test_json/test_pandas.py [hunk 2]
```diff
@@ -815,11 +815,39 @@ def test_mixed_timedelta_datetime(self):
...
         assert_frame_equal(expected, result, check_index_type=False)
 
+    def test_default_handler_indirect(self):
+        from pandas.io.json import dumps
+
+        def default(obj):
+            if isinstance(obj, complex):
+                # Series + OrderedDict to get predictable serialisation result
+                return Series(OrderedDict([
+                    ('mathjs', 'Complex'),
```
Comment (jreback) [inline pandas/io/tests/test_json/test_pandas.py]: is the default handler allowed to raise? if so do we ignore it or raise?

File: pandas/io/tests/test_json/test_pandas.py [hunk 3]
```diff
@@ -815,11 +815,39 @@ def test_mixed_timedelta_datetime(self):
...
     def test_default_handler(self):
         value = object()
-        frame = DataFrame({'a': ['a', value]})
-        expected = frame.applymap(str)
+        frame = DataFrame({'a': [7, value]})
+        expected = DataFrame({'a': [7, str(value)]})
         result = pd.read_json(frame.to_json(default_handler=str))
         assert_frame_equal(expected, result, check_index_type=False)
 
+    def test_default_handler_indirect(self):
```
Comment (jreback) [inline pandas/io/tests/test_json/test_pandas.py]: can you add a small section on the default handler in ``io.rst``

**Comment by jreback** [comment]
ok so other PR merged, you can rebase

File: pandas/io/tests/test_json/test_pandas.py (hunk 3 above)
Comment (Komnomnomnom) [inline pandas/io/tests/test_json/test_pandas.py]: It's already covered [here](http://pandas-docs.github.io/pandas-docs-travis/io.html#fallback-behavior) but I brought it up to date.

File: pandas/io/tests/test_json/test_pandas.py (hunk 2 above)
Comment (Komnomnomnom) [inline pandas/io/tests/test_json/test_pandas.py]: Yes, it's [already tested](https://github.com/pydata/pandas/blob/master/pandas/io/tests/json/test_pandas.py#L817). The exception is bubbled up

File: doc/source/whatsnew/v0.18.1.txt (hunk 1 above)
Comment (Komnomnomnom) [inline doc/source/whatsnew/v0.18.1.txt]: fixed

**Comment by Komnomnomnom** [comment]
FYI windows test https://ci.appveyor.com/project/Komnomnomnom/pandas/build/1.0.4 . Thanks for the tip!

**Comment by jreback** [comment]
@Komnomnomnom gr8!. note that appveyor is REALLY SLOW!

File: doc/source/io.rst [hunk 4]
```diff
@@ -1501,45 +1501,34 @@ Fallback Behavior
...
-is to specify a ``default_handler``. For example ``timedelta`` can cause
-problems:
+- if an object is unsupported it will attempt the following:
 
-.. ipython:: python
-   :suppress:
 
-   from datetime import timedelta
-   dftd = DataFrame([timedelta(23), timedelta(seconds=5), 42])
+  * check if the object has defined a ``toDict`` method and call it.
```
Comment (jreback) [inline doc/source/io.rst]: where did ``toDict`` come from?

Comment (Komnomnomnom) [inline doc/source/io.rst]: [ujson](https://github.com/esnme/ultrajson/blob/master/python/objToJSON.c#L825). Has been there since the beginning

Comment (jreback) [inline doc/source/io.rst]: oh, ok, so its 'standard'

Comment (Komnomnomnom) [inline doc/source/io.rst]: I am :+1: on deprecating / removing `toDict` and the object traversal behaviour.

Comment (Komnomnomnom) [inline doc/source/io.rst]: afaik ujson is the only library that supports `toDict` and fallback to object traversal (i.e. serialising obj.__dict__)

Comment (jreback) [inline doc/source/io.rst]: sure, go ahead and deprecate and use ``obj.__dict__``

Comment (Komnomnomnom) [inline doc/source/io.rst]: To be clear I'm suggesting deprecating `toDict` **and** the fallback to `obj.__dict__`

So `default_handler` would be the only fallback behaviour (and possibly a future [`default_enc`](https://github.com/pydata/pandas/pull/12739#issuecomment-214761708))

Comment (jreback) [inline doc/source/io.rst]: oh, I c.

for sure deprecate ``toDict()``.

is there something problematic with ``obj.__dict__``? e.g. maybe as a fallback to ``default_handler`` (which you would use first, or if ``None`` then use ``obj.__dict__``).

Comment (Komnomnomnom) [inline doc/source/io.rst]: The current implementation uses `dir(obj)` and just encodes 'public' attrs. It works ok sometimes.

I don't have strong feelings on removing it but In my experience unless the objects are very simple it's generally not what you want and  it's also quite brittle (`OverflowError`s are common).

Comment (jreback) [inline doc/source/io.rst]: ok, with deprecating non-supported things and just using ``default_handler`` that's what its there for. in theory using ``__json__`` might be ok, but not really sure what is the 'standard'.

**Comment by Komnomnomnom** [comment]
@jreback build clean (apart from msgpack & pickle again)
https://travis-ci.org/Komnomnomnom/pandas/builds/126098431

also windows is ok:
https://ci.appveyor.com/project/Komnomnomnom/pandas/build/1.0.8

**Comment by jreback** [comment]
just remember how to fix the tag issue.

on your local master
``git push yourorigin master --tags``

tags come from your fork (and not from the clone, pandas)

next time you branch you should be good

**Comment by jreback** [comment]
thanks!

make an issue about the deprecations for 0.19.0 pls.

**Pull request closed without merging**